### PR TITLE
understand-tidb: add rbo section

### LINKED ISSUE #59: Write down the first version of Rule-based Optimization section
We don't have a file for it yet. May we create one on writing.

cc @yudongusa

### PR DESCRIPTION
**Pull request by time-and-fate**
### What issue does this PR solve?

- close #59

### What is changed:

Add the Rule-based Optimization section.

### CODE REVIEW

**Review by zhangyangyu — APPROVED** [review]

**Comment by zhangyangyu** [comment]
@tisonkun @yudongusa Do you think it's ready to merge?

File: src/understand-tidb/rbo.md [hunk 1]
```diff
@@ -1 +1,523 @@
...
+|   └─Selection_11(Probe)         | 1.00     | cop[tikv] |                      | or(eq(test2.t2.a, 1), eq(test2.t2.a, 2))                                                                                                                                                                                 |
+|     └─TableRowIDScan_10         | 1.00     | cop[tikv] | table:t2             | keep order:false, stats:pseudo                                                                                                                                                                                           |
++---------------------------------+----------+-----------+----------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+```
+
+### Outer Join Elimination
+
+This rule finds and tries to eliminate `Join`. Specifically, it removes the `Join` and its inner side sub-plan tree.
```
Comment (winoros) [inline src/understand-tidb/rbo.md:312]: We should specify that we are eliminating the OUTER JOIN, not INNER JOIN.

File: src/understand-tidb/rbo.md [hunk 2]
```diff
@@ -1 +1,523 @@
...
+
+[Pushing `Aggregation` across `Join`](https://github.com/pingcap/tidb/blob/94e30df8e2d8ba2a1a26f153f40067ba3acd78eb/planner/core/rule_aggregation_push_down.go#L405) is the most complicated case of them. The aggregate functions are separated into left and right sides and we try to push them to the left and right side of `Join` respectively. There are many requirements to make this transformation. For example, the join type should be among `InnerJoin`, `LeftOuterJoin` and `RightOuterJoin`. Only specific types of aggregate functions can be pushed down. And when we try to push aggregate functions down to one side of the `Join`, there can't be `count()` and `sum()` functions in the other side. If all requirements are met, we can generate and push down `Aggregation`. The new `Aggregation` is also transformed and different from the original `Aggregation`. For example, the columns in the join conditions should be added into `GroupByItems`.
```
Comment (winoros) [inline src/understand-tidb/rbo.md:358]: Tell the reader that this is not always good so it's not opened by default yet.

File: src/understand-tidb/rbo.md [hunk 3]
```diff
@@ -1 +1,523 @@
...
+|     └─TableReader_17               | 8000.00  | root      |               | data:HashAgg_12                                                                              |
+|       └─HashAgg_12                 | 8000.00  | cop[tikv] |               | group by:test2.t1.a, funcs:max(test2.t1.b)->Column#11                                        |
+|         └─TableFullScan_15         | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                               |
++------------------------------------+----------+-----------+---------------+----------------------------------------------------------------------------------------------+
+```
+
+[Pushing `Aggregation` across `Projection`](https://github.com/pingcap/tidb/blob/94e30df8e2d8ba2a1a26f153f40067ba3acd78eb/planner/core/rule_aggregation_push_down.go#L436) is rather simple. It directly moves the calculation in the `Projection` into `Aggregation`. Then the `Projection` can be removed. And there will be only one `Aggregation`.
```
Comment (winoros) [inline src/understand-tidb/rbo.md:385]: Tell the reader that this one is always beneficial.

File: src/understand-tidb/rbo.md [hunk 4]
```diff
@@ -1 +1,523 @@
...
+
+[Pushing `Aggregation` across `UnionAll` and `PartitionUnionAll`](https://github.com/pingcap/tidb/blob/94e30df8e2d8ba2a1a26f153f40067ba3acd78eb/planner/core/rule_aggregation_push_down.go#L475-L485) share the same logic. It's similar to the `LogicalJoin` case. There will be some checks. If all requirements are met. We can generate and push down a new `LogicalAggregation` across `UnionAll` or `PartitionUnionAll`. Note that the original `Aggregation` may also be modified here.
```
Comment (winoros) [inline src/understand-tidb/rbo.md:387]: Tell the reader that only push across the partition union can be treated as beneficial.

File: src/understand-tidb/rbo.md [hunk 5]
```diff
@@ -1 +1,523 @@
...
+
+## Rules
+
+### Column Pruning
+
+This is a very fundamental optimization. It will prune unneeded columns for each operator. 
+
+This main logic is in `PruneColumns(parentUsedCols []*expression.Column) error` method of the `LogicalPlan` interface. It traverses the plan tree from top to bottom. Each operator receives which columns are used by the parent operator, then uses this information to prune unneeded columns from itself (different kinds of operator would have different behaviors), then collect and pass columns needed by itself to its children.
+
+### Decorrelation
```
Comment (winoros) [inline src/understand-tidb/rbo.md:37]: Tell the reader that the decorrelate is not always beneficial.

File: src/understand-tidb/rbo.md [hunk 6]
```diff
@@ -1 +1,523 @@
...
+| ├─TableReader_15(Build)      | 9990.00  | root      |               | data:Selection_14                             |
+| │ └─Selection_14             | 9990.00  | cop[tikv] |               | not(isnull(test2.t1.a))                       |
+| │   └─TableFullScan_13       | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                |
+| └─TableReader_12(Probe)      | 9990.00  | root      |               | data:Selection_11                             |
+|   └─Selection_11             | 9990.00  | cop[tikv] |               | not(isnull(test2.t.a))                        |
+|     └─TableFullScan_10       | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                |
++------------------------------+----------+-----------+---------------+-----------------------------------------------+
+```
+
+Second, we will also try to derive some extra conditions from the existing predicates or try to add `NOT NULL` when possible. This enables us to push more predicates down.
```
Comment (winoros) [inline src/understand-tidb/rbo.md:289]:
```suggestion
Second, we will also try to derive some extra conditions that are the common conditions from the existing OR predicates or try to add `NOT NULL` when possible. This enables us to push more predicates down.
```

**Comment by time-and-fate** [comment]
@winoros Updated. PTAL~

**Pull request merged**